src/stake/stake_keys.rs

```diff
@@ -0,0 +1,170 @@
+
+            let start = cmp::min(range.start, bounds.start);
+            let end = cmp::max(range.end, bounds.end);
+            let bounds = Range { start, end };
+
+            // Not using the entry manipulation because there is no inherent optimization in doing
+            // that for the cost of manually replacing the hash entry bucket of the key.
+            let _was_removed = self.keys.remove(&range).is_some();
+
+            debug_assert!(_was_removed);
+
+            self.keys.insert(bounds, stake);
+
+            return Ok(());
+        }
+
+        // Extend is not possible because either the key or value differs - intersection is forbidden
+        // from this point
+        if let Some(range) = self
+            .keys
+            .keys()
+            .find(|range| range.contains(&bounds.start) || range.contains(&bounds.end))
+        {
+            return Err(Error::DuplicatedStake {
+                height: range.start,
+                valid_before: range.end,
+            });
+        }
+
+        self.keys.insert(bounds, stake);
+
+        Ok(())
+    }
+
+    /// Remove all stake entries
+    pub(super) fn clear(&mut self) {
+        self.keys.clear();
+    }
+
+    /// Return a stake for a given height
+    pub(super) fn fetch(&self, height: Height) -> Option<&Stake> {
+        self.keys
+            .iter()
+            .filter_map(|(range, stake)| range.contains(&height).then(|| stake))
+            .next()
+    }
+
+    /// Remove all entries with the provided key
+    pub(super) fn purge_key(&mut self, key: &PublicKey) {
+        self.keys.retain(|_, stake| &stake.key != key);
+    }
+
+    /// Iter the ranges and stakes
+    pub(super) fn iter(&self) -> impl Iterator<Item = (&Range<Height>, &Stake)> {
+        self.keys.iter()
+    }
+}
+
+fn normalize_range<R>(bounds: R) -> Range<Height>
+where
+    R: RangeBounds<Height>,
+{
+    let start = match bounds.start_bound() {
+        Bound::Included(s) => *s,
+        Bound::Excluded(s) if s == &Height::MIN => Height::MIN,
+        Bound::Excluded(s) => s.saturating_add(1),
+        Bound::Unbounded => Height::MIN,
+    };
+
+    let end = match bounds.end_bound() {
+        Bound::Included(e) => e.saturating_add(1),
+        Bound::Excluded(e) => *e,
+        Bound::Unbounded => Height::MAX,
+    };
+
+    Range { start, end }
+}
+
+#[test]
+#[cfg(feature = "std")]
+fn stake_keys_intersect_and_merge() {
+    use fuel_crypto::SecretKey;
+    use rand::rngs::StdRng;
+    use rand::{Rng, SeedableRng};
+
+    let rng = &mut StdRng::seed_from_u64(0xbeef);
+
+    // keys
+    let a = SecretKey::random(rng).public_key();
+    let b = SecretKey::random(rng).public_key();
+
+    // values
+    let x = rng.gen();
+    let y = rng.gen();
+
+    let ax = Stake { key: a, value: x };
+    let ay = Stake { key: a, value: y };
+    let by = Stake { key: b, value: y };
+
+    let mut keys = StakeKeys::default();
+    keys.add_stake_range(0..2, ax).expect("no intersect");
```

Comment:
Do we expect to be using range syntax like this when programmatically adding these stake ranges in normal usage? I've found it can be difficult to create range objects without using this special syntax.

Reply:
This is as close as it gets to rust stdlib.

I don't see any major complications there since any range implemented in rust also implements RangeBounds. This means we can:

```rust
pool.stake(validator, .., stake); // stake up to infinity
pool.stake(validator, start..end, stake); // stake from start to end
pool.stake(validator, first.., stake); // stake from first to infinity
```

Reply:
As long as this interface lets us properly specify the boundary cases (e.g exclusive vs inclusive etc) then I think this should be fine.